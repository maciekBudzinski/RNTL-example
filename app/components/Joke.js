import React, {useState, useEffect} from 'react';
import {View, Text, StyleSheet} from 'react-native';
import axios from 'axios';

const useChuckNorrisJokeFetch = () => {
  const [joke, setJoke] = useState('');

  const getTextFromServer = async () => {
    const response = await axios.get('https://api.chucknorris.io/jokes/random');
    setJoke(response.data.value);
  };

  useEffect(() => {
    getTextFromServer();
  }, []);

  return joke;
};

const Joke = () => {
  const jokeText = useChuckNorrisJokeFetch();

  return (
    <View style={styles.wrapper}>
      <Text style={styles.text}>{jokeText}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    padding: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  text: {
    fontSize: 22,
  },
});

export default Joke;
