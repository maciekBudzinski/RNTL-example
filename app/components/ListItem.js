import React, { useCallback, useEffect } from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';

const ListItem = ({
  name,
  id,
  deleteItem,
  handleUnmount = () => console.log('unmount'),
  handleNameChange = itemName => console.log(itemName),
}) => {
  const handleDeletePress = useCallback(() => {
    deleteItem(id);
  }, [deleteItem, id]);

  useEffect(() => handleUnmount());

  useEffect(() => {
    handleNameChange(name);
  }, [handleNameChange, name]);

  return (
    <View testID="wrapper" style={styles.wrapper}>
      <Text style={styles.text}>{name}</Text>
      <View style={styles.buttonsContainer}>
        <TouchableOpacity onPress={handleDeletePress}>
          <Text style={styles.deleteItem}>Delete</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    padding: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  text: {
    fontSize: 22,
  },
  buttonsContainer: {
    alignItems: 'center',
  },
  deleteItem: {
    color: 'red',
  },

});

export default ListItem;
