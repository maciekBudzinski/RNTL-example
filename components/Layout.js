import React from 'react';
import {View, SafeAreaView, StyleSheet} from 'react-native';

const Layout = ({children}) => (
  <SafeAreaView style={styles.safeArea}>
    <View style={styles.wrapper}>
      {children}
    </View>
  </SafeAreaView>
);

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
  },
  wrapper: {
    padding: 16,
    flex: 1,
  },
});

export default Layout;
