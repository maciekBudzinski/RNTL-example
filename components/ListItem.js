import React, {useCallback} from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';

const ListItem = ({name, id, deleteItem}) => {
  const handleDeletePress = useCallback(()=> {
    deleteItem(id);
  },[deleteItem, id]);

  return (
    <View testID="wrapper" style={styles.wrapper}>
      <Text style={styles.text}>{name}</Text>
      <View style={styles.buttonsContainer}>
        <TouchableOpacity onPress={handleDeletePress}>
          <Text style={styles.deleteItem}>Delete</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    padding: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  text: {
    fontSize: 22,
  },
  buttonsContainer: {
    alignItems: 'center',
  },
  deleteItem: {
    color: 'red',
  },

});

export default ListItem;
