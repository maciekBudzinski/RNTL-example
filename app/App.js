import React, { useState, useCallback } from 'react';
import {
  Alert,
  Text,
  TextInput,
  Button,
  FlatList,
} from 'react-native';
import { Layout, ListItem, Joke } from 'components';

const showEmptyInputAlert = () => {
  Alert.alert('Error', 'Can\'t add empty item', [{
    text: 'Ok',
  }]);
};

const App = ({ defaultItems = [] }) => {
  const [itemName, setItemName] = useState('');
  const [items, setItems] = useState(defaultItems);

  const handleInputChange = useCallback((text) => {
    setItemName(text);
  }, []);

  const addItem = useCallback((isAsync = false) => {
    if (itemName) {
      setItems([...items, { id: Date.now(), name: (isAsync ? 'async' : '') + itemName }]);
      setItemName('');
    }
    else {
      showEmptyInputAlert();
    }
  }, [itemName, items]);

  const addAsyncItem = () => (
    new Promise((resolve) => {
      setTimeout(() => {
        addItem(true);
        resolve();
      }, Math.random() * 1000);
    })
  );

  const deleteItem = useCallback((itemId) => {
    setItems(items.filter(({ id }) => id !== itemId));
  }, [items]);

  const isListEmpty = items.length === 0;

  return (
    <Layout>
      <Text testID="testID">Todo App</Text>
      <TextInput placeholder="Todo name" value={itemName} onChangeText={handleInputChange} />
      <Button title="Add item" onPress={addItem} />
      <Button title="Add async item" onPress={addAsyncItem} />
      {isListEmpty && <Text>No items</Text>}
      <FlatList
        data={items}
        keyExtractor={({ id }) => id.toString()}
        renderItem={({ item }) =>
          (
            <ListItem
              deleteItem={deleteItem}
              {...item}
            />)}
      />
      <Joke />
    </Layout>
  );
};

export default App;
