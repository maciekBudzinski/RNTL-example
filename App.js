import React, {useState, useCallback} from 'react';
import {
  Alert,
  Text,
  TextInput,
  Button,
} from 'react-native';

import Layout from './components/Layout';
import ListItem from './components/ListItem';

const showEmptyInputAlert = () => {
  Alert.alert('Error', 'Can\'t add empty item', [{
    text: 'Ok',
  }]);
};

const App = () => {
  const [itemName, setItemName] = useState('');
  const [items, setItems] = useState([]);

  const handleInputChange = useCallback((text)=> {
    setItemName(text);
  }, []);

  const addItem = useCallback(()=> {
    if (itemName) {
      setItems([...items, {id: Date.now(), name: itemName}]);
      setItemName('');
    }
    else {
      showEmptyInputAlert();
    }
  }, [itemName, items]);


  const deleteItem = useCallback((itemId)=> {
    setItems(items.filter(({id})=> id !== itemId));
  }, [items]);

  const isListEmpty = items.length === 0;
  return (
    <Layout>
      <Text>Todo App</Text>
      <TextInput placeholder="Todo name" value={itemName} onChangeText={handleInputChange} />
      <Button title="Add item" onPress={addItem} />
      {isListEmpty && <Text>No items</Text>}
      {items.map(({name, id})=> <ListItem key={id} name={name} id={id} deleteItem={deleteItem} />)}
    </Layout>
  );
};

export default App;
