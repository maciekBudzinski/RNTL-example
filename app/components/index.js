import Layout from './Layout';
import ListItem from './ListItem';
import Joke from './Joke';

export {
  Layout,
  ListItem,
  Joke,
};
