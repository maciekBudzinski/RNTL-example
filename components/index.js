import Layout from './Layout';
import ListItem from './ListItem';

export {
  Layout,
  ListItem,
};
