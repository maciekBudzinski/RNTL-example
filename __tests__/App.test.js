import React from 'react';
import {Button, TouchableOpacity} from 'react-native';
import {ListItem} from 'components';
import App from '../App';
import { render, debug, fireEvent, waitForElement } from 'react-native-testing-library';

describe('App', () => {
  it('Should match snapshot', () => {
    const wrapper = render(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should add item correctly', ()=>{
    const newTodoName = 'Test todo';

    const wrapper = render(<App />);

    const textInput = wrapper.getByPlaceholder('Todo name');
    fireEvent.changeText(textInput, newTodoName);

    const addButton = wrapper.getByType(Button);
    fireEvent.press(addButton);

    expect(wrapper.getByType(ListItem)).toBeTruthy();
    expect(wrapper.getByText(newTodoName)).toBeTruthy();
  });

  it('Should delete items', ()=> {
    const props = {
      defaultItems : [{
        id: 123,
        name: 'Todo 123',
      },
      {
        id: 456,
        name: 'Todo 456',
      },
      ],
    };

    const wrapper = render(<App {...props} />);
    const deleteButtons = wrapper.getAllByText('Delete');
    fireEvent.press(deleteButtons[0]);
    expect(wrapper.getAllByType(ListItem).length).toBe(1);
    expect(wrapper.getAllByText('Todo 456').length).toBe(1);
  });
});
