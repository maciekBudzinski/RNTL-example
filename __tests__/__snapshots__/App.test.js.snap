// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 16,
      }
    }
  >
    <Text
      testID="testID"
    >
      Todo App
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Todo name"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add item
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add async item
        </Text>
      </View>
    </View>
    <Text>
      No items
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 8,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 22,
          }
        }
      >
        
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
